File: Data Structures & Algorithms/combination-target-sum/submission-1.cpp
class Solution {
    void backtrack(int index, int sum, vector<int>& curr, vector<int>& nums,
                   vector<vector<int>>& output, int n, int target) {
        if (sum == target){
            output.push_back(curr);
            return ;}
        if (sum > target || index==n)
            return;
            if (index > 0 && nums[index] == nums[index - 1] && sum == 0) return;

            curr.push_back(nums[index]);
            // Recurse to form the next combination
            sum += nums[index];
            backtrack(index, sum, curr, nums, output, n, target);
            // Backtrack and remove the last element
            curr.pop_back();
            sum -= nums[index];
                        backtrack(index+1, sum, curr, nums, output, n, target);

       
    }

public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
    sort(nums.begin(), nums.end());
    vector<vector<int>> res;
    vector<int> curr;
    backtrack(0, 0, curr, nums, res, nums.size(), target);
    return res;
}

};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        nums.sort()
        def backtrack(i,subset):
          
            if sum(subset)== target:
                res.append(subset.copy())
                return
            if sum(subset)> target or i>=len(nums):
                return 
            subset.append(nums[i])
            backtrack(i+1,subset)
            subset.pop()
            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            backtrack(i + 1, subset)
            
        backtrack(0,[])
        return res